skimage/transform/registration.py

```diff
@@ -0,0 +1,185 @@
+    inverse : bool, optional
+        Whether to return the inverse transform, which converts coordinates
+        in the reference space to coordinates in the target space. For
+        technical reasons, this is the transform expected by
+        ``scipy.ndimage.affine_transform`` to map the target image to the
+        reference space.
+    level_callback : callable, optional
+        If given, this function is called once per pyramid level with a tuple
+        containing the current downsampled image, transformation matrix, and
+        cost as the argument. This is useful for debugging or for plotting
+        intermediate results during the iterative process.
+    method : string
+        Method of minimization.  See ``scipy.optimize.minimize`` for available
+        options.
+    **kwargs : keyword arguments
+        Keyword arguments passed through to ``scipy.optimize.minimize``
+
+    Returns
+    -------
+    matrix : array
+        A transformation matrix used to obtain a new image.
+        ndi.affine_transform(target, matrix) will align your target image.
+
+    Example
+    -------
+    >>> from skimage.data import astronaut
+    >>> reference_image = astronaut()[..., 1]
+    >>> r = -0.12  # radians
+    >>> c, s = np.cos(r), np.sin(r)
+    >>> matrix_transform = np.array([[c, -s, 0], [s, c, 50], [0, 0, 1]])
+    >>> target_image = ndi.affine_transform(reference_image, matrix_transform)
+    >>> matrix = register_affine(reference_image, target_image)
+    >>> registered_target = ndi.affine_transform(target_image, matrix)
+    """
+
+    # ignore the channels if present
+    ndim = reference.ndim if not multichannel else reference.ndim - 1
+    if ndim == 0:
+        raise ValueError(
+            'Input images must have at least 1 spatial dimension.')
+
+    min_dim = min(reference.shape[:ndim])
+    nlevels = int(np.floor(np.log2(min_dim) - np.log2(minimum_size)))
+
+    pyramid_ref = pyramid_gaussian(reference, max_layer=nlevels,
+                                   multichannel=multichannel)
+    pyramid_tgt = pyramid_gaussian(target, max_layer=nlevels,
+                                   multichannel=multichannel)
+    image_pairs = reversed(list(zip(pyramid_ref, pyramid_tgt)))
+    parameter_vector = _matrix_to_parameter_vector(np.identity(ndim + 1))
+
+    for ref, tgt in image_pairs:
+        def _cost(param):
+            transformation = _parameter_vector_to_matrix(param, ndim)
+            if not multichannel:
+                transformed = ndi.affine_transform(tgt, transformation,
+                                                   order=1)
+            else:
+                transformed = np.zeros_like(tgt)
+                for ch in range(tgt.shape[-1]):
+                    ndi.affine_transform(tgt[..., ch], transformation,
+                                         order=1, output=transformed[..., ch])
+            return cost(ref, transformed)
```

Comment:
Why not define the parameter cost function `_param_cost(param, image0, image1, cost, multichannel)` outside the `register_affine` function definition block 
```python

def _param_cost(param, image0, image1, cost, multichannel):
    transformation = _parameter_vector_to_matrix(param, ndim)
    if not multichannel:
       transformed = ndi.affine_transform(image1, transformation,
                                           order=1)
    else:
        transformed = np.zeros_like(image1)
        for ch in range(image1.shape[-1]):
            ndi.affine_transform(image1[..., ch], transformation,
                                 order=1, output=transformed[..., ch])
    return cost(image0, transformed)

```
and use
```suggestion
        _cost = functools.partial(_param_cost, image0=ref, image1=tgt, cost=cost,
                                  multichannel=multichannel)
```

Reply:
I'm more used to closures, but I like your approach! I'll give it a go. =)